### PR TITLE
[CORDA-1763]: Add node CLI option for validating configuration.

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-1763

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 1]
```diff
@@ -57,56 +57,56 @@ abstract class NodeCliCommand(alias: String, description: String, val startup: N
...
 open class NodeStartupCli : CordaCliWrapper("corda", "Runs a Corda Node") {
     open val startup = NodeStartup()
+    @Mixin
+    val cmdLineOptions = NodeCmdLineOptions()
 
-    private val networkCacheCli by lazy { ClearNetworkCacheCli(startup) }
-    private val justGenerateNodeInfoCli by lazy { GenerateNodeInfoCli(startup) }
-    private val justGenerateRpcSslCertsCli by lazy { GenerateRpcSslCertsCli(startup) }
-    private val initialRegistrationCli by lazy { InitialRegistrationCli(startup) }
+    private val networkCache by lazy { ClearNetworkCacheCli(startup) }
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: Keep the `Cli` suffix to make it clear in uses further down.

File: node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+internal class ValidateConfigurationCli(private val cmdLineOptions: SharedNodeCmdLineOptions) : CliWrapperBase("validate-configuration", "Validates the configuration without starting the node.") {
+
+    private companion object {
+
+        private val logger = loggerFor<ValidateConfigurationCli>()
+    }
+
+    override fun runProgram(): Int {
+
```
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]: As an aside, I actually find this much harder to see the structure of the code when scrolling as I can't see the differences between properties/functions.

File: node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt [hunk 3]
```diff
@@ -168,6 +162,11 @@ open class NodeStartup : NodeStartupLogging {
         return ExitCodes.SUCCESS
     }
 
+    private fun logConfigurationErrors(errors: Set<Exception>) {
+
+        logger.error("Invalid node configuration. Errors were:${System.lineSeparator()}${errors.asSequence().map(Exception::message).joinToString(System.lineSeparator())}")
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/NodeStartup.kt]: What if there's only 1 error ;) `Errors were:` vs `Error was:`

File: node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import net.corda.cliutils.CliWrapperBase
+import net.corda.cliutils.ExitCodes
+import net.corda.common.validation.internal.Validated
+import net.corda.common.validation.internal.Validated.Companion.invalid
+import net.corda.common.validation.internal.Validated.Companion.valid
+import net.corda.core.utilities.loggerFor
+import net.corda.node.SharedNodeCmdLineOptions
+import net.corda.node.services.config.NodeConfiguration
+
+internal class ValidateConfigurationCli(private val cmdLineOptions: SharedNodeCmdLineOptions) : CliWrapperBase("validate-configuration", "Validates the configuration without starting the node.") {
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]: _Imperative mood._ Validates => Validate

File: node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    private companion object {
+
+        private val logger = loggerFor<ValidateConfigurationCli>()
+    }
+
+    override fun runProgram(): Int {
+
+        val configuration = cmdLineOptions.nodeConfiguration()
+        if (configuration.isInvalid) {
+            logger.error("Invalid node configuration. Errors were:${System.lineSeparator()}${configuration.errorsDescription()}")
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]: What about when there's a single error.

File: docs/source/changelog.rst [hunk 6]
```diff
@@ -7,6 +7,8 @@ release, see :doc:`upgrade-notes`.
 Unreleased
 ----------
 
+* Added new "validate-configuration" sub-command to `corda.jar`, allowing to validate the actual node configuration without starting the node.
```
Comment (tlil) [inline docs/source/changelog.rst]: Imperative mood vs past tense.

File: docs/source/running-a-node.rst [hunk 7]
```diff
@@ -66,6 +66,8 @@ Sub-commands
 
 ``initial-registration``: Starts initial node registration with the compatibility zone to obtain a certificate from the Doorman.
 
+``validate-configuration``: Validates the actual configuration without starting the node.
```
Comment (tlil) [inline docs/source/running-a-node.rst]: Should tidy up these as we're currently using a mix of imperative mood and present tense.

File: node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]: I'm not improving the messages as part of this story, just exposing the existing situation through a sub-command as well. These comments will be addressed by using the specification.

About duplicated error message for the invalid char, I'll have a look.

File: docs/source/running-a-node.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (sollecitom) [reply docs/source/running-a-node.rst]: Done, thanks :)

File: node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (anthonykeenan) [reply node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]: If you don't want to inherit from `NodeCliCommand` then add this line to the object:

    `@Mixin
    val cmdLineOptions = SharedNodeCmdLineOptions()`

Which will add the options you need.

Also, if you don't override `initLogging()` then by default you will end up with a log file in the root dir.

File: node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]:
```
If you run java -jar corda-4.0-SNAPSHOT.jar validate-configuration with no config file you get:
[ERROR] 14:44:43+0100 [main] subcommands.ValidateConfigurationCli.runProgram - Invalid node configuration. Errors were: C:\ak\1763\node.conf: java.io.FileNotFoundException: C:\ak\1763\node.conf (The system cannot find the file specified)
```

Fixed ^^^

[the remark this answers is not in the document]
Comment (mikehearn) [reply node/src/main/kotlin/net/corda/node/internal/subcommands/ValidateConfigurationCli.kt]: Actually the code style is to not have such blank lines. Please avoid adding them. When we upgrade to Kotlin 1.3 and do a reformatting pass I suspect they'll be automatically removed.

**Pull request merged**